src/libraries/Microsoft.PowerFx.Core/Public/Values/CollectionTableValue.cs

```diff
@@ -223,14 +245,34 @@ protected override async Task<DValue<RecordValue>> PatchCoreAsync(RecordValue ba
         /// <remarks>A derived class may override if there's a more efficient way to find the match than by linear scan.</remarks>
         protected virtual async Task<RecordValue> FindAsync(RecordValue baseRecord, CancellationToken cancellationToken, bool mutationCopy = false)
         {
-            if (this is IMutationCopy && mutationCopy)
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (baseRecord.TryGetPrimaryKey(out string basePrimaryKey))

```

Comment:
I added the following lines to MatchesAsync but, while this technically works (all UTs are Ok), we lose the performance benefit in this case.
We end up in `RecordValue copyRecord = (RecordValue)record.MaybeShallowCopy();` part of the code which is VERY slow

            if (baseRecord.TryGetPrimaryKey(out string basePrimaryKey) && currentRecord.TryGetPrimaryKey(out string recordPrimaryKeyValue) && basePrimaryKey == recordPrimaryKeyValue)
            {
                return true;
            }

The shallow copy scans all fields of the Entity, following relationships... hence many network calls

Reply:
1. By skipping maybeShallowCopy, I'm wondering if there is actually a correctness issue here - we may be missing some test cases (see comments in test file)?

2. "The shallow copy scans all fields of the Entity, following relationships... hence many network calls."     A shallow copy most certainly should not be making network calls; this may be hitting some of the other DV bugs that greg filed.

Reply:
This is resolved now